File: tests/fixtures/sample-python-project-negative/services/api_gateway/services/user_service.py
import os
import requests
from shared.utils.formatters import format_user

USER_SERVICE_URL = os.environ.get("USER_SERVICE_URL", "http://localhost:3001")


# VIOLATION: style/deterministic/docstring-completeness
class UserService:
    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/no-self-use
    def find_all(self) -> list:
        # VIOLATION: reliability/deterministic/http-call-no-timeout
        response = requests.get(f"{USER_SERVICE_URL}/users")
        return [format_user(u) for u in response.json()]

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/no-self-use
    def find_by_id(self, user_id: str) -> dict | None:
        # VIOLATION: reliability/deterministic/http-call-no-timeout
        response = requests.get(f"{USER_SERVICE_URL}/users/{user_id}")
        return format_user(response.json())

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/no-self-use
    def create(self, data: dict) -> dict:
        # VIOLATION: reliability/deterministic/http-call-no-timeout
        response = requests.post(f"{USER_SERVICE_URL}/users", json=data)
        return format_user(response.json())

    # VIOLATION: style/deterministic/docstring-completeness
    # VIOLATION: code-quality/deterministic/no-self-use
    def delete(self, user_id: str) -> None:
        # VIOLATION: reliability/deterministic/http-call-no-timeout
        requests.delete(f"{USER_SERVICE_URL}/users/{user_id}")

File: tests/fixtures/sample-js-project-negative/services/api-gateway/src/services/user.service.ts
import axios from 'axios';
import { formatUser } from '@sample/shared-utils';

// NOTE: code-quality/deterministic/env-in-library-code — skipped for non-packages files
const USER_SERVICE_URL = process.env.USER_SERVICE_URL || 'http://localhost:3001';

export class UserService {
  // VIOLATION: code-quality/deterministic/missing-return-type
  // VIOLATION: code-quality/deterministic/static-method-candidate
  async findAll() {
    // VIOLATION: reliability/deterministic/http-call-no-timeout
    const response = await axios.get(`${USER_SERVICE_URL}/users`);
    return response.data.map(formatUser);
  }

  // VIOLATION: code-quality/deterministic/missing-return-type
  // VIOLATION: code-quality/deterministic/static-method-candidate
  async findById(id: string) {
    // VIOLATION: reliability/deterministic/http-call-no-timeout
    const response = await axios.get(`${USER_SERVICE_URL}/users/${id}`);
    return formatUser(response.data);
  }

  // VIOLATION: code-quality/deterministic/missing-return-type
  // VIOLATION: code-quality/deterministic/static-method-candidate
  async create(data: { name: string; email: string }) {
    // VIOLATION: reliability/deterministic/http-call-no-timeout
    const response = await axios.post(`${USER_SERVICE_URL}/users`, data);
    return formatUser(response.data);
  }

  // VIOLATION: code-quality/deterministic/missing-return-type
  // VIOLATION: code-quality/deterministic/static-method-candidate
  async delete(id: string) {
    // VIOLATION: reliability/deterministic/http-call-no-timeout
    await axios.delete(`${USER_SERVICE_URL}/users/${id}`);
  }
}
